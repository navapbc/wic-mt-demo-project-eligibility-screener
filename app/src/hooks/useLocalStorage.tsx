import { SessionData } from '@customTypes/common'
import { useState } from 'react'

// Custom hook to persist state across page refresh
export default function useLocalStorage(
  key: string,
  initialValue: SessionData
) {
  // Pass initial state function to useState so logic is only executed once
  const [storedValue, setStoredValue] = useState<SessionData>(() => {
    if (typeof window === 'undefined') {
      return initialValue
    }
    try {
      // Get from local storage by key
      const item: string | null = window.sessionStorage.getItem(key)

      // TODO: refactor type casting
      const state: SessionData = item
        ? (JSON.parse(item) as SessionData)
        : initialValue

      return state
    } catch (error) {
      // TODO: error handling
      console.log(error)
      return initialValue
    }
  })
  // Return a wrapped version of useState's setter function that persists the new value to sessionStorage.
  const setValue = (value: SessionData) => {
    try {
      // Allow value to be a function so we have same API as useState

      // TODO: refactor type casting
      const valueToStore: SessionData =
        value instanceof Function ? (value(storedValue) as SessionData) : value
      // Save state
      setStoredValue(valueToStore)
      // Save to local storage
      if (typeof window !== 'undefined') {
        window.sessionStorage.setItem(key, JSON.stringify(valueToStore))
      }
    } catch (error) {
      console.log(error)
    }
  }
  return [storedValue, setValue]
}
