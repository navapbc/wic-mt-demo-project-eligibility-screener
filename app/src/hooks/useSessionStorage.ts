import { useState } from 'react'

import { SessionData } from '@src/types'

// Custom hook to persist state across page refresh
export default function useSessionStorage(
  key: string,
  initialValue: SessionData
) {
  // Pass initial state function to useState so logic is only executed once
  const [storedValue, setStoredValue] = useState<SessionData>(() => {
    if (typeof window === 'undefined') {
      return initialValue
    }
    try {
      // Get from local storage by key
      const item: string | null = window.sessionStorage.getItem(key)

      const state: SessionData = item
        ? (JSON.parse(item) as SessionData)
        : initialValue

      return state
    } catch (error) {
      // @TODO: Production-readiness: handle more robust logging and error handling.
      console.log(error)
      return initialValue
    }
  })
  // Return a wrapped version of useState's setter function that persists the new value to sessionStorage.
  const setValue = (value: SessionData) => {
    try {
      // Allow value to be a function so we have same API as useState
      const valueToStore: SessionData =
        value instanceof Function ? (value(storedValue) as SessionData) : value
      // Save state
      setStoredValue(valueToStore)
      // Save to local storage
      if (typeof window !== 'undefined') {
        window.sessionStorage.setItem(key, JSON.stringify(valueToStore))
      }
    } catch (error) {
      console.log(error)
    }
  }
  return [storedValue, setValue]
}

export function clearSessionStorage(key: string) {
  if (typeof window === 'undefined') {
    // @TODO: Production-readiness: handle more robust logging and error handling.
    console.log('Not in client context')
  } else {
    window.sessionStorage.removeItem(key)
  }
}
