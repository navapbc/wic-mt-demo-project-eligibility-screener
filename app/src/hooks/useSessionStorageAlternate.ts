// Approach taken by Nick Scialli
// See https://typeofnan.dev/using-session-storage-in-react-with-hooks

import { useEffect, useState } from 'react'

function getSessionStorageOrDefault(key, defaultValue) {
  const stored = sessionStorage.getItem(key)
  if (!stored) {
    return defaultValue
  }
  return JSON.parse(stored)
}

export default function useSessionStorage(key, defaultValue) {
  const [value, setValue] = useState(
    getSessionStorageOrDefault(key, defaultValue)
  )

  useEffect(() => {
    sessionStorage.setItem(key, JSON.stringify(value))
  }, [key, value])

  return [value, setValue]
}
