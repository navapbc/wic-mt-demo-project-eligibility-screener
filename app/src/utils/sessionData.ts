import type {
  ChooseClinicData,
  ContactData,
  EligibilityData,
  IncomeData,
  SessionData,
} from '@src/types'

export const initialChooseClinicData: ChooseClinicData = {
  clinic: undefined,
  zipCode: '',
}

export const initialEligibilityData: EligibilityData = {
  residential: '',
  categorical: [],
  previouslyEnrolled: '',
  adjunctive: [],
}

export const initialContactData: ContactData = {
  firstName: '',
  lastName: '',
  phone: '',
  comments: '',
}

export const initialIncomeData: IncomeData = {
  householdSize: '',
}

export const initialSessionData: SessionData = {
  chooseClinic: initialChooseClinicData,
  contact: initialContactData,
  eligibility: initialEligibilityData,
  income: initialIncomeData,
}
