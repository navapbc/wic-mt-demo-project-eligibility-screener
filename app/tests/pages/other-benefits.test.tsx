import { render, screen } from '@testing-library/react'
import singletonRouter from 'next/router'

import OtherBenefits from '@pages/other-benefits'

import * as sessionModule from '@src/hooks/useSessionStorage'
import { initialSessionData } from '@utils/sessionData'

import { setMockSession, setup } from '../helpers/setup'
import {
  testAccessibility,
  testBackLink,
  testSnapshot,
} from '../helpers/sharedTests'

/**
 * Test setup
 */

const route = '/other-benefits'
const backRoute = '/eligibility'
const forwardRoute = '/'

/**
 * Begin tests
 */

it('should match full page snapshot', () => {
  const { mockSession } = setup(route)
  testSnapshot(
    <OtherBenefits
      sessionKey="mockSessionKey"
      setSession={setMockSession}
      session={mockSession}
      backRoute={backRoute}
      forwardRoute={forwardRoute}
    />
  )
})

it('should pass accessibility scan', async () => {
  const { mockSession } = setup(route)
  await testAccessibility(
    <OtherBenefits
      sessionKey="mockSessionKey"
      setSession={setMockSession}
      session={mockSession}
      backRoute={backRoute}
      forwardRoute={forwardRoute}
    />
  )
})

it('should have a back link that matches the backRoute', () => {
  const { mockSession } = setup(route)
  testBackLink(
    <OtherBenefits
      sessionKey="mockSessionKey"
      setSession={setMockSession}
      session={mockSession}
      backRoute={backRoute}
      forwardRoute={forwardRoute}
    />,
    backRoute
  )
})

it('should clear the session and redirect when the button is clicked', async () => {
  const { mockSession, user } = setup(route)
  // Spy on clearSessionStorage() to make sure it gets called.
  const spy = jest.spyOn(sessionModule, 'clearSessionStorage')
  render(
    <OtherBenefits
      sessionKey="mockSessionKey"
      setSession={setMockSession}
      session={mockSession}
      backRoute={backRoute}
      forwardRoute={forwardRoute}
    />
  )
  const button = screen.getByRole('button', { name: /Return/i })
  await user.click(button)

  expect(setMockSession).toHaveBeenCalledWith(initialSessionData)
  expect(singletonRouter).toMatchObject({ asPath: forwardRoute })
  expect(spy).toHaveBeenCalled()
})
