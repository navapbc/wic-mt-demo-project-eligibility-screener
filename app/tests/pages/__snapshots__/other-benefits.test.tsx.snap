// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match full page snapshot 1`] = `
[
  <a
    className="usa-link"
    href="/eligibility"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onTouchStart={[Function]}
  >
    Back
  </a>,
  <h1>
    Your eligibility
  </h1>,
  <h2>
    Given your responses, you are likely not eligible for WIC in Montana.
  </h2>,
  <p>
    We recommend visiting 
    <a
      className="usa-link usa-link--external"
      href="https://dphhs.mt.gov/Assistance"
      rel="noopener noreferrer"
      target="_blank"
    >
      https://dphhs.mt.gov/Assistance
    </a>
     to learn about other assistance in Montana.
  </p>,
  <p>
    If you are not currently living or working in Montana, visit 
    <a
      className="usa-link usa-link--external"
      href="https://www.signupwic.com/"
      rel="noopener noreferrer"
      target="_blank"
    >
      SignUpWIC.com
    </a>
     to find a clinic near you.
  </p>,
  <button
    className="usa-button usa-button--small display-block margin-top-6"
    onClick={[Function]}
  >
    Return to start page
  </button>,
]
`;
